File: PA2/B/Submission.java
import java.util.*;
import java.io.*;
import java.lang.*;

public class Submission {
    public static void main(String[] args) {
        FastIO io = new FastIO();
        int TC = Integer.parseInt(io.nextLine());
        for (int i=0; i<TC; i++) {
            long N = io.nextLong();
            long A = io.nextLong();
            long B = io.nextLong();
            long C = io.nextLong();
            long X = io.nextLong();
            long Y = io.nextLong();

            long[] lst = new long[(int) N];
            long previous = A;
            lst[0] = A;

            for (int j=1; j<N; j++) {
                long current = (A + ((previous % C) * B)) % C * A;
                lst[(int) j] = current;
                previous = current;
            }

            long result;
            if (N % 2 == 0) {
                int index1 = (int) (N/2 - 1);
                int index2 = (int) (N/2);
                long first = find(lst, index1);
                long second = find(lst, index2);
                result = (first + second) / 2;
            } else {
                int index = (((int) N)/2);
                result = find(lst, index);
            }
            io.println(result);
        }
        io.close();
    }
    static long find(long[] lst, int index) {
        if (lst.length == 1) {
            return lst[0];
        } else {
            int size = lst.length;
            long[] auxiliary = new long[size];
            int smaller = 0;
            int larger = size-1;
            int chosen = (int) (Math.floor(Math.random() * size));
            long chosenNum = lst[chosen];
            for (int i=0; i<size; i++) {
                if (i != chosen) {
                    long toCompare = lst[i];
                    if (toCompare <= chosenNum) {
                        auxiliary[smaller] = toCompare;
                        smaller++;
                    } else {
                        auxiliary[larger] = toCompare;
                        larger--;
                    }
                }
            }
            if (smaller == index) {
                return chosenNum;
            } else if (smaller < index) {
                long[] newList = new long[size - smaller - 1];
                for (int j=0; j<newList.length; j++) {
                    newList[j] = lst[smaller+j+1];
                }
                return find(newList, index-smaller-1);
            } else {
                long[] newList = new long[smaller];
                for (int j=0; j<newList.length; j++) {
                    newList[j] = lst[j];
                }
                return find(newList, index);
            }
        }
    }
}

// io class
class FastIO extends PrintWriter {
    BufferedReader br;
    StringTokenizer st;

    public FastIO() {
        super(new BufferedOutputStream(System.out));
        br = new BufferedReader(new InputStreamReader(System.in));
    }

    String next() {
        while (st == null || !st.hasMoreElements()) {
            try {
                st = new StringTokenizer(br.readLine());
            } catch (IOException e) {
                e.printStackTrace();
            }
        }
        return st.nextToken();
    }

    int nextInt() {
        return Integer.parseInt(next());
    }

    long nextLong() {
        return Long.parseLong(next());
    }

    double nextDouble() {
        return Double.parseDouble(next());
    }

    String nextLine() {
        String str = "";
        try {
            str = br.readLine();
        } catch (IOException e) {
            e.printStackTrace();
        }
        return str;
    }
}

File: PA2/B/Submission.cpp
#include <bits/stdc++.h>
#include <iostream>
#include <string>
#include <vector>
#include <algorithm>
#include <limits>

using namespace std;

// io class
class FastIO {
    public:
        FastIO() {
            ios_base::sync_with_stdio(false);
            cin.tie(NULL);
        }

        string next() {
            cin >> buffer;
            return buffer;
        }

        int nextInt() {
            return stoi(next());
        }

        long long nextLong() {
            return stoll(next());
        }

        double nextDouble() {
            return stod(next());
        }

        string nextLine() {
            getline(cin, buffer);
            return buffer;
        }

        void println(long long x) {
            cout << x << "\n";
        }

        void close() {
            cout.flush();
            cin.clear();
            cin.ignore(numeric_limits<streamsize>::max(), '\n');
        }

    private:
        string buffer;
};

int main() {
    FastIO io;
    int TC = io.nextInt();
    for (int i = 0; i < TC; i++) {
        int N = io.nextInt();
        long A = io.nextLong();
        long B = io.nextLong();
        long C = io.nextLong();
        long X = io.nextLong();
        long Y = io.nextLong();

        long lst[N];

        long previous = A;
        lst[0] = A;
        for (int j = 1; j < N; j++) {
            long current = ((A + (previous % C) * B) % C) * A;
            lst[j] = current;
            previous = current;
        }

        if (N % 2 == 1) {
            int index = N / 2;
            std::nth_element(lst, lst + index, lst + N);
            io.println(lst[index]);
        } else {
            int index = N / 2;
            std::nth_element(lst, lst + index - 1, lst + N);
            long long smaller = lst[index - 1];
            std::nth_element(lst, lst + index, lst + N);
            long long larger = lst[index];
            long long raw = smaller + larger;
            if (raw % 2 == 0) {
                io.println(raw / 2);
            } else {
                long d = raw / 2;
                cout<< d <<".5\n";
            }
        }
    }
    io.close();
    return 0;
}